### PR TITLE
santio/fix/multi connection pool

### PR DESCRIPTION
**Pull request by SantioMC**
- **fix: multi connection pool**

### CODE REVIEW

File: src/main/kotlin/gg/ingot/iron/pool/MultiConnectionPool.kt [hunk 1]
```diff
@@ -110,15 +74,53 @@ class MultiConnectionPool(
...
-        openConnections.set(0)
+    override fun close(force: Boolean) {
+        logger.trace("Closing connection pool, force: {}", force)
+        if (force) {
+            coroutineScope.cancel()
+            pool.forEach(Connection::close)
+        }
+    }
+
+    private fun push(connection: Connection) {
```
Comment (DebitCardz) [inline src/main/kotlin/gg/ingot/iron/pool/MultiConnectionPool.kt:85]: docs

File: src/main/kotlin/gg/ingot/iron/pool/MultiConnectionPool.kt [hunk 2]
```diff
@@ -110,15 +74,53 @@ class MultiConnectionPool(
...
+        if (!offered) {
+            logger.trace("Connection pool is full, closing connection.")
+            connection.close()
+        }
+
+        logger.trace("Connection task finished, open connections: {}", openConnections.get())
+        openConnections.decrementAndGet()
+    }
+
+    private fun pull(): Connection? {
```
Comment (DebitCardz) [inline src/main/kotlin/gg/ingot/iron/pool/MultiConnectionPool.kt:102]: docs

File: src/main/kotlin/gg/ingot/iron/pool/MultiConnectionPool.kt [hunk 3]
```diff
@@ -110,15 +74,53 @@ class MultiConnectionPool(
...
+        }
+
+        val offered = pool.offer(connection, iron.settings.connectionPollTimeout.inWholeMilliseconds, TimeUnit.MILLISECONDS)
+        if (!offered) {
+            logger.trace("Connection pool is full, closing connection.")
+            connection.close()
+        }
+
+        logger.trace("Connection task finished, open connections: {}", openConnections.get())
+        openConnections.decrementAndGet()
```
Comment (DebitCardz) [inline src/main/kotlin/gg/ingot/iron/pool/MultiConnectionPool.kt:99]: can we set a variable of the output of `AtomicInteger#decrementAndGet` and use that for the log instead?

File: src/main/kotlin/gg/ingot/iron/pool/MultiConnectionPool.kt [hunk 4]
```diff
@@ -110,15 +74,53 @@ class MultiConnectionPool(
...
+
+        if (openConnections.get() >= iron.settings.maximumConnections) {
+            logger.trace("Maximum connections reached, waiting for a connection to be released.")
+            return pool.poll(iron.settings.connectionPollTimeout.inWholeMilliseconds, TimeUnit.MILLISECONDS)
+        }
+
+        val connection = pool.poll() ?: createConnection()
+        openConnections.incrementAndGet()
+
+        logger.trace("Acquired a connection from the pool, open connections: {}", openConnections.get())
```
Comment (DebitCardz) [inline src/main/kotlin/gg/ingot/iron/pool/MultiConnectionPool.kt:116]: can we also set this to a variable for logging? just for consistency.

**Review by DebitCardz — CHANGES_REQUESTED** [review]

**Comment by SantioMC** [comment]
Closing in favor of #21

**Pull request closed without merging**